### PR TITLE
Use a common validation module for both frontend and backend

### LINKED ISSUE #135: Use a common validator configuration for both frontend and backend
Since I will also do client-side validation and notify the user when information is invalid, I have set up a common validator file for both frontend and backend, at `misc/validate.ts`.

As of now, it is

```ts
export const SITE = {
    nameIsValid: (name: string) => isSlug(name),
    domainIsValid: (domain: string) =>
        isURL(domain, { require_protocol: true, protocols: ["http", "https"] }),
};
```

Please also migrate all of your data validation to this file (so that I can use it too).

While we are at it, I suggest the following validation:

  * Site name should be `isSlug()` since we are using it in the URL
  * Since we will use the domain in the [`frame-ancestor` field](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-ancestors), I think `str === "*" || isFQDN(str, { allow_wildcard: true })` is better suited (so yes, current `SITE.domainIsValid` is incorrect):

    ```
    [Log] google.com – true
    [Log] https://google.com – false
    [Log] my.google.com – true
    [Log] https://my.google.com – false
    [Log] google.com/hello – false
    [Log] https://google.com/hello – false
    [Log] my.google.com/hello – false
    [Log] https://my.google.com/hello – false
    [Log] http://google.com – false
    [Log] random://google.com – false
    [Log] *.google.com – true
    [Log] https://*.google.com – false
    [Log] google.com.vn.a.very.long.domain.name.com.vn – true
    ```

    For the case of `*`, we will simply remove the `frame-ancestors` field.

### PR DESCRIPTION
**Pull request by joulev**
This will close #135.

### CODE REVIEW

**Comment by joulev** [comment]
Hold on, I will write a couple of unit tests for this before this is ready for merging.

**Comment by VietAnh1010** [comment]
Will review this tomorrow.

File: jest.client.config.js [hunk 1]
```diff
@@ -4,7 +4,7 @@ const createJestConfig = nextJest({ dir: "./" });
 
 const customJestConfig = {
   // setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
-  testRegex: "/__tests__/client/.*\\.[jt]sx?$",
+  testRegex: "/__tests__/(client|misc)/.*\\.[jt]sx?$",
```
Comment (VietAnh1010) [inline jest.client.config.js:7]: Maybe seperate `misc` to be a seperate test suite, as both client and server depend on it?

File: misc/validate.ts [hunk 2]
```diff
@@ -5,11 +5,33 @@
  * validation logic on both ends. So that we can make change easily whenever we need to.
  */
 // Import this way for the sake of tree-shaking
+import isFQDN from "validator/lib/isFQDN";
 import isSlug from "validator/lib/isSlug";
 import isURL from "validator/lib/isURL";
 
+export const USER = {
+    nameIsValid: (name: string) => name.length > 0,
```
Comment (VietAnh1010) [inline misc/validate.ts:13]: Should be changed to `displayNameIsValid`, for a more consistent naming convention.

Comment (VietAnh1010) [reply misc/validate.ts:13]: For JS `name.length` shoule be enough. But if you want to be explicit, then this is fine.

File: misc/validate.ts [hunk 3]
```diff
@@ -5,11 +5,33 @@
...
-        isURL(domain, { require_protocol: true, protocols: ["http", "https"] }),
+    domainIsValid: (domain: string) => domain === "*" || isFQDN(domain, { allow_wildcard: true }),
+    // server-only
+    uidIsValid: (uid: string) => uid.length > 0 && !uid.includes(" "),
+    // obsolete?
+    iconURLIsValid: (iconURL: string) => isURL(iconURL, { require_protocol: true }),
+};
+
+export const PAGE = {
+    nameIsValid: (title: string) => title.length > 0,
```
Comment (VietAnh1010) [inline misc/validate.ts:26]: Will be changed to `titleIsValid`

**Review by VietAnh1010 — COMMENTED** [review]
I left some comment about some possible changes. They are not too significant, just make the code looks more consistent.

All look good. We will merge this after merging #140.

File: jest.client.config.js (hunk 1 above)
Comment (joulev) [reply jest.client.config.js:7]: I think I will keep it this way. The client test suite is "bare" Jest (super fast) which is sufficient for the validator too, while the server test suite will have Firebase emulator (not necessary). Of course in the future if the validator (or anything in `misc`) is complicated enough we will have a separate test suite for it.

**Comment by VietAnh1010** [comment]
> This branch cannot be rebased due to conflicts

What the heck?!

**Pull request merged**